piet-cairo/src/text.rs

```diff
@@ -210,235 +357,149 @@ impl TextLayout for CairoTextLayout {
+             */
+            self.pango_layout.set_width(-1);
+        }
 
-    // Now that we know it's not beginning or end, begin binary search.
-    // Iterative style
-    let mut left = start;
-    let mut right = end;
-    loop {
-        // pick halfway point
-        let middle = left + ((right - left) / 2);
-
-        let grapheme_bounds = match get_grapheme_boundaries(font, text, middle) {
-            Some(bounds) => bounds,
-            None => return HitTestPoint::default(),
-        };
+        let mut line_metrics = Vec::new();
+        let mut y_distance = 0.;
+        let mut widest_logical_width = 0;
+        for line_index in 0..self.pango_layout.get_line_count() {
+            if let Some(line) = self.pango_layout.get_line_readonly(line_index) {
+                /*
+                 * NOTE: These values are not currently exposed so we need to get them
+                 * manually. It's kinda sucky I know
+                 *
+                 * TODO(ForLoveOfCats): Submit a PR to gtk-rs to expose these values
+                 */
+                let (start_offset, end_offset) = unsafe {
+                    let raw_line = line.to_glib_none();
+
+                    let start_offset = (*raw_line.0).start_index as usize;
+                    let length = (*raw_line.0).length as usize;
+
+                    (start_offset, start_offset + length)
+                };
 
-        if let Some(hit) = point_x_in_grapheme(point.x, &grapheme_bounds) {
-            return hit;
-        }
+                //Pango likes to give us the line range *without* the newline char(s)
+                let end_offset = {
+                    let mut end_offset = end_offset;
+                    loop {
+                        let next_byte = self.text.as_bytes().get(end_offset);
+                        match next_byte {
+                            Some(b'\r') | Some(b'\n') => {
+                                end_offset += 1;
+                            }
+
+                            _ => break end_offset,
+                        }
+                    }
+                };
 
-        // since it's not a hit, check if closer to start or finish
-        // and move the appropriate search boundary
-        if point.x < grapheme_bounds.leading {
-            right = middle;
-        } else if point.x > grapheme_bounds.trailing {
-            left = middle + 1;
-        } else {
-            unreachable!("hit_test_point conditional is exhaustive");
+                let ink_rect = line.get_extents().0;
+                let logical_rect = line.get_extents().1;
+
+                if logical_rect.width > widest_logical_width {
+                    widest_logical_width = logical_rect.width;
+                }
+
+                let line_text = &self.text[start_offset..end_offset];
+                let trimmed_len = line_text.trim_end().len();
+                let trailing_whitespace = line_text[trimmed_len..].chars().count();
```

Comment:
this should be in utf-8 code-units, so just `.len()` instead of `.chars()`.

Reply:
Also if there's not test for this I'd want to really double check that what we get back is inclusive of trailing whitespace and not of trailing newlines.

Reply:
Oops, that's a little embarrassing :sweat_smile: 

So trailing whitespace should not include newline characters, that does make sense

Reply:
My understanding is that the return value *should* return up to one trailing newline sequence. This behavior is derived from [DWrite](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/ns-dwrite-dwrite_line_metrics), and, while we can make arguments about whether it would be better to include or exclude, it is important to be consistent.

Reply:
I swapped the `.chars().count()` for a `.len()` as the `end_offset` logic above will ensure that the first newline sequence for that line are included in the trailing whitespace